fbw-common/src/wasm/terronnd/src/navigationdisplay/display.h

```diff
@@ -113,14 +114,37 @@ class Display : public DisplayBase {
     this->_frameData->defineArea(side == DisplaySide::Left ? FrameDataLeftName : FrameDataRightName);
     this->_frameData->requestArea(SIMCONNECT_CLIENT_DATA_PERIOD_ON_SET);
     this->_frameData->setOnChangeCallback([=]() {
-      this->destroyImage();
-
       if (!this->_ignoreNextFrame && this->_configuration.terrainActive) {
-        this->_nanovgImage =
-            nvgCreateImageMem(this->_context, 0, this->_frameData->data().data(), static_cast<int>(this->_frameBufferSize));
         if (this->_nanovgImage == 0) {
-          std::cerr << "TERR ON ND: Unable to decode the image from the stream" << std::endl;
+          // If we don't have an image yet, create one
+          this->_nanovgImage = nvgCreateImageMem(this->_context, 0, this->_frameData->data().data(), static_cast<int>(this->_frameBufferSize));
+          if (this->_nanovgImage == 0) {
+            fprintf(stderr, "TERR ON ND: Unable to decode the image from the stream. Reason: %s\n", stbi_failure_reason());
+          }
+
+          return;
+        }
+
+        // Otherwise, decode the PNG manually and update the existing image
+        int decodedWidth, decodedHeight;
+        uint8_t* decodedImage = stbi_load_from_memory(this->_frameData->data().data(), static_cast<int>(this->_frameBufferSize), &decodedWidth, &decodedHeight, nullptr, 4);
+        if (decodedImage == nullptr) {
+          fprintf(stderr, "TERR ON ND: Unable to decode the image from the stream. Reason: %s\n", stbi_failure_reason());
+          return;
         }
+
+        int width, height;
+        nvgImageSize(this->_context, this->_nanovgImage, &width, &height);
+
+        if (decodedWidth != width || decodedHeight != height) {
```

Comment:
Does the image not need a cleanup in this case as well?

Reply:
Good catch, fixed. If only we had defer..